# 🐼 Pandas 기초와 실습


## Pandas란 무엇인가?

Pandas는 Python의 데이터 분석 및 조작을 위한 핵심 라이브러리입니다.  
엑셀의 스프레드시트와 테이블 형식의 데이터 (DataFrame)을 쉽고 직관적으로 다룰 수 있게 해주는 도구입니다.

데이터 분석에서 Pandas가 필수인 이유
- **다양한 데이터 형태 지원**: 숫자, 문자, 날짜 등 혼합된 데이터 처리 가능
- **직관적인 표 형태**: 엑셀과 유사한 방식으로 데이터 조작
- **강력한 데이터 처리**: 대용량 데이터도 빠르고 효율적으로 처리
- **유연한 데이터 변환**: 필터링, 정렬, 그룹화 등 다양한 분석 작업 지원

의료 분야에서는 환자 데이터, 검사 결과, 진료 기록 등의 복합적인 의료 정보를 분석하고 관리하는 데 사용됩니다.


### Pandas 연습 사이트: https://www.machinelearningplus.com/python/101-pandas-exercises-python/



### NumPy vs Pandas

둘 다 데이터 분석에서 핵심 라이브러리입니다.

NumPy는 숫자 계산용(수학·배열 연산), Pandas는 표 형태 데이터 분석용(엑셀처럼 행·열 처리)입니다.

**pandas는 numpy의 기반 위에서 만들어진 고급 데이터 분석 도구**로, numpy만으로 분석하기 힘든 데이터 연산을 해줍니다.




### Pandas 라이브러리 import

Pandas 사용을 위해 필요한 라이브러리를 불러오고 정상 작동을 확인합니다.


In [2]:
# Pandas 라이브러리 불러오기
import pandas as pd  # 데이터 분석용 라이브러리

# 버전 확인
print(f"Pandas 버전: {pd.__version__}")

Pandas 버전: 2.2.2


## 데이터 구조 생성하기

Pandas의 두 가지 핵심 데이터 구조인 Series와 DataFrame을 이해하고 다양한 방법으로 생성해봅니다.


### Series

Series는 1차원 배열로, 인덱스(라벨)와 값의 쌍으로 구성됩니다.

**Series 특징**
- **구조**: 인덱스(라벨) + 값의 1차원 데이터
- **비유**: 엑셀에서 한 개 열을 선택한 것과 동일
- **활용**: 각 데이터에 의미있는 이름을 붙여 관리 가능

```python
# Series 생성 문법
# pd.Series(데이터, index=라벨목록)
```


In [8]:
# Series 예시: 환자별 수축기 혈압 정보
patient_bp = pd.Series(
    [120, 130, 140, 110, 125],  # 혈압 수치 (값)
    index=["김철수", "이영희", "박민수", "최지은", "정도영"],
)  # 환자 이름 (인덱스)

print("환자별 수축기 혈압 (Series)")
patient_bp

환자별 수축기 혈압 (Series)


,0
김철수,120
이영희,130
박민수,140
최지은,110
정도영,125


In [9]:
print(f"Series 타입: {type(patient_bp)}")
print(f"김철수의 혈압: {patient_bp['김철수']} mmHg")  # 인덱스로 특정 값 접근

Series 타입: <class 'pandas.core.series.Series'>
김철수의 혈압: 120 mmHg


### DataFrame

DataFrame은 2차원 표 형태로, 행과 열로 구성된 구조화된 데이터를 저장합니다.

**DataFrame 특징**
- **구조**: 행(index)과 열(columns)로 구성된 2차원 테이블
- **비유**: 엑셀 스프레드시트와 완전히 동일한 개념
- **활용**: 여러 종류의 정보를 하나의 표로 관리
- **관계**: 여러 개의 Series를 모아놓은 구조

행(row)과 컬럼(column) 은 다음과 같은 의미를 가집니다.

- 행(row): ```하나의 개체(관측 대상, record)``` 를 의미합니다.  
예) 환자 1명, 진료 기록 1건

- 컬럼(column): ```데이터의 속성(특징, feature)``` 을 의미합니다.  
예) 나이, 성별, 혈압, 혈당, BMI 등

#### DataFrame 생성 2가지 방법

DataFrame을 생성하는 주요 방법을 알아봅시다.


- **딕셔너리 방식**: 컬럼 중심, 직관적이고 읽기 쉬움
- **2차원 배열 방식**: 행별 레코드, 코드가 간결하고 속도가 딕셔너리 방식보다  빠름



In [10]:
# DataFrame 생성용 공통 변수
patient_names = ["김철수", "이영희", "박민수", "최지은"]
columns = ["나이", "혈압", "혈당"]

##### 1. 딕셔너리로 생성 (가장 일반적)


In [11]:
# 1. 딕셔너리로 생성 - 컬럼 중심 데이터
patient_dict = {
    "나이": [45, 67, 32, 29],  # 열 단위로 데이터 구성
    "혈압": [120, 140, 110, 115],
    "혈당": [95, 180, 88, 92],
}
df1 = pd.DataFrame(patient_dict, index=patient_names)

print("딕셔너리로 생성한 DataFrame:")
df1

딕셔너리로 생성한 DataFrame:


,나이,혈압,혈당
김철수,45,120,95
이영희,67,140,180
박민수,32,110,88
최지은,29,115,92


##### 2. 2차원 배열로 생성


In [12]:
# 2. 딕셔너리 리스트로 생성 - 행별 딕셔너리
patient_records = [
    [45, 120, 95],  # 김철수 (행)
    [67, 140, 180],  # 이영희
    [32, 110, 88],  # 박민수
    [29, 115, 92],  # 최지은
]
df2 = pd.DataFrame(patient_records, index=patient_names, columns=columns)

print("딕셔너리 리스트로 생성한 DataFrame: ")

df2

딕셔너리 리스트로 생성한 DataFrame: 


,나이,혈압,혈당
김철수,45,120,95
이영희,67,140,180
박민수,32,110,88
최지은,29,115,92


In [13]:
# DataFrame 예시: 10명 환자의 종합 건강 정보
patient_data = {
    "환자ID": [
        "P001",
        "P002",
        "P003",
        "P004",
        "P005",
        "P006",
        "P007",
        "P008",
        "P009",
        "P010",
    ],  # 환자 식별번호
    "이름": [
        "김철수",
        "이영희",
        "박민수",
        "최지은",
        "정도영",
        "장미래",
        "오성진",
        "한소희",
        "윤태완",
        "송지원",
    ],  # 환자 성명
    "나이": [45, 67, 32, 29, 12, 41, 35, 102, 28, 63],  # 만 나이
    "성별": ["남", "여", "남", "여", "남", "여", "남", "여", "남", "여"],  # 성별 구분
    "키": [175, 162, 180, 158, 170, 165, 172, 160, 178, 155],  # 키 (cm)
    "체중": [70, 58, 85, 52, 75, 60, 68, 65, 82, 48],  # 체중 (kg)
    "수축기혈압": [120, 140, 110, 115, 135, 125, 130, 145, 118, 120],  # 수축기 혈압 (mmHg)
    "이완기혈압": [80, 90, 70, 75, 85, 82, 88, 95, 78, 92],  # 이완기 혈압 (mmHg)
    "혈당": [95, 180, 88, 92, 110, 102, 98, 156, 89, 134],  # 공복혈당 (mg/dL)
    "흡연": [
        "비흡연",
        "금연",
        "비흡연",
        "비흡연",
        "흡연",
        "비흡연",
        "흡연",
        "금연",
        "비흡연",
        "흡연",
    ],  # 흡연 상태
    "운동": [
        "규칙적",
        "가끔",
        "안함",
        "규칙적",
        "가끔",
        "규칙적",
        "안함",
        "가끔",
        "규칙적",
        "안함",
    ],  # 운동 습관
}

health_df = pd.DataFrame(patient_data)  # 딕셔너리를 DataFrame으로 변환

print("환자 건강 정보 (DataFrame)")
health_df

환자 건강 정보 (DataFrame)


,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동
0,P001,김철수,45,남,175,70,120,80,95,비흡연,규칙적
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔
2,P003,박민수,32,남,180,85,110,70,88,비흡연,안함
3,P004,최지은,29,여,158,52,115,75,92,비흡연,규칙적
4,P005,정도영,12,남,170,75,135,85,110,흡연,가끔
5,P006,장미래,41,여,165,60,125,82,102,비흡연,규칙적
6,P007,오성진,35,남,172,68,130,88,98,흡연,안함
7,P008,한소희,102,여,160,65,145,95,156,금연,가끔
8,P009,윤태완,28,남,178,82,118,78,89,비흡연,규칙적
9,P010,송지원,63,여,155,48,120,92,134,흡연,안함


In [ ]:
print(f"타입: {type(health_df)}")
print(f"데이터 크기: {health_df.shape})")  # (행 수, 열 수) 튜플 반환

타입: <class 'pandas.core.frame.DataFrame'>
데이터 크기: (10, 11))


### DataFrame vs Series

DataFrame에서 한 개 컬럼을 선택하면 Series가 되고, 여러 컬럼을 선택하면 DataFrame이 됩니다.

**핵심 문법**
- `df['컬럼명']` → Series (1차원)
- `df[['컬럼명1', '컬럼명2']]` → DataFrame (2차원)  
- DataFrame = 여러 개의 Series로 구성된 구조


In [ ]:
print("DataFrame의 한 컬럼은 Series")
age_series = health_df["나이"]  # 한 개 컬럼 선택 → Series 반환

print(f"나이 컬럼 타입: {type(age_series)}")
print()

age_series

DataFrame의 한 컬럼은 Series
나이 컬럼 타입: <class 'pandas.core.series.Series'>



,나이
0,45
1,67
2,32
3,29
4,12
5,41
6,35
7,102
8,28
9,63


In [ ]:
print("DataFrame의 여러 컬럼은 DataFrame")
basic_info_df = health_df[["이름", "나이", "성별"]]  # 여러 컬럼 선택 → DataFrame 반환

print(f"기본정보 타입: {type(basic_info_df)}")
print()

basic_info_df

DataFrame의 여러 컬럼은 DataFrame
기본정보 타입: <class 'pandas.core.frame.DataFrame'>



,이름,나이,성별
0,김철수,45,남
1,이영희,67,여
2,박민수,32,남
3,최지은,29,여
4,정도영,12,남
5,장미래,41,여
6,오성진,35,남
7,한소희,102,여
8,윤태완,28,남
9,송지원,63,여


### 🤔 실습 문제

다음과 같은 데이터 구조를 만들어보세요.
1. **환자 체중 정보**를 담은 **Series** 생성 (5명의 환자 이름을 인덱스로 사용)
2. **병원 직원 정보**를 담은 **DataFrame** 생성 (이름, 부서, 경력 컬럼 포함)


In [ ]:
# 여기에 작성하세요


In [ ]:
# @title 정답
# 1. 환자 체중 정보 Series 생성
patient_weight = pd.Series(
    [65, 72, 58, 80, 70],  # 체중 데이터 (값)
    index=["환자A", "환자B", "환자C", "환자D", "환자E"],
)  # 환자 이름 (인덱스)

print("환자 체중 정보 (Series)")
print(patient_weight)
print()

# 2. 병원 직원 정보 DataFrame 생성
staff_data = {
    "이름": ["김의사", "이간호사", "박기사", "최약사"],  # 직원 이름
    "부서": ["내과", "간호부", "영상의학과", "약제부"],  # 소속 부서
    "경력": [5, 3, 8, 2],  # 근무 경력 (년)
}

staff_df = pd.DataFrame(staff_data)  # 딕셔너리를 DataFrame으로 변환

print("병원 직원 정보 (DataFrame)")
print(staff_df)

환자 체중 정보 (Series)
환자A    65
환자B    72
환자C    58
환자D    80
환자E    70
dtype: int64

병원 직원 정보 (DataFrame)
     이름     부서  경력
0   김의사     내과   5
1  이간호사    간호부   3
2   박기사  영상의학과   8
3   최약사    약제부   2


## CSV 파일 읽기/쓰기

CSV(Comma-Separated Values)는 데이터 분석의 핵심 파일 형식입니다.

**실무에서 데이터 분석 시 압도적으로 많이 사용되는 방식**으로, 대부분의 데이터 사이언티스들이 일상적으로 활용하는 표준 형식입니다.

**CSV가 데이터 분석에서 널리 사용되는 이유**
- **범용성**: Excel, Google Sheets, 데이터베이스 등 어디서나 읽고 쓸 수 있음
- **호환성**: 운영체제나 소프트웨어에 관계없이 동일하게 동작
- **경량성**: 텍스트 기반으로 용량이 작고 전송 속도가 빠름
- **가독성**: 사람이 직접 읽고 편집할 수 있는 구조
- **표준화**: 웹 크롤링, API 응답, 데이터베이스 내보내기의 기본 형식


**데이터 분석 워크플로우**
```
원시 데이터 수집 → CSV로 저장 → Pandas로 분석 → 결과를 다시 CSV로 내보내기
```


### CSV 파일로 저장하기

DataFrame을 CSV 파일로 저장하여 다른 프로그램에서도 사용할 수 있게 만듭니다.


In [14]:
# CSV 파일로 저장
health_df.to_csv(
    "clinic_health_data.csv",  # 저장할 파일명
    index=False,  # 행 번호(인덱스)는 저장하지 않음
    encoding="utf-8-sig",
)  # 한글 깨짐 방지용 인코딩

print("CSV 파일 저장 완료: 'clinic_health_data.csv'")

CSV 파일 저장 완료: 'clinic_health_data.csv'


### CSV 파일 불러오기

저장된 CSV 파일을 다시 읽어서 DataFrame으로 변환하여 분석에 활용합니다.


In [15]:
# CSV 파일에서 데이터 읽기
df = pd.read_csv("clinic_health_data.csv")  # CSV 파일을 DataFrame으로 불러오기

print(f"불러온 데이터 크기: {df.shape}")
print("첫 3행 미리보기:")
df.head(3)  # 처음 3행만 출력하여 데이터 구조 확인

불러온 데이터 크기: (10, 11)
첫 3행 미리보기:


,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동
0,P001,김철수,45,남,175,70,120,80,95,비흡연,규칙적
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔
2,P003,박민수,32,남,180,85,110,70,88,비흡연,안함


## 데이터 탐색 및 구조 속성 파악

데이터 분석의 시작 단계로, 데이터의 전체적인 구조와 특성을 파악하는 탐색적 데이터 분석(EDA) 과정입니다.


### 기본 데이터 탐색

데이터를 처음 받았을 때 가장 먼저 확인해야 할 기본 정보들을 추출하는 방법입니다.

**주요 메서드**
- `len(df)`: 데이터의 행의 갯수 확인
- `head(n)`: 데이터의 처음 n행을 확인하여 데이터 구조 파악
- `tail(n)`: 데이터의 마지막 n행을 확인하여 데이터 끝부분 상태 점검  
- `info()`: 데이터 타입, 결측치, 메모리 사용량 등 전반적 정보 확인
- `describe()`: 숫자형 데이터의 기본 통계량(평균, 표준편차, 최솟값, 최댓값 등) 확인

```python
# 데이터 탐색 문법
# len(df)         # 행 갯수
# df.head(n=5)    # 처음 n행
# df.tail(n=5)    # 마지막 n행  
# df.info()       # 전반적 정보
# df.describe()   # 기본 통계량
```


In [17]:
print(f"총 {len(df)}명의 환자 데이터")

총 10명의 환자 데이터


In [ ]:
# 처음과 끝 데이터 확인
print("처음 5명:")
df.head()  # 기본값은 5행

처음 5명:


,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동
0,P001,김철수,45,남,175,70,120,80,95,비흡연,규칙적
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔
2,P003,박민수,32,남,180,85,110,70,88,비흡연,안함
3,P004,최지은,29,여,158,52,115,75,92,비흡연,규칙적
4,P005,정도영,12,남,170,75,135,85,110,흡연,가끔


In [ ]:
print("마지막 3명:")
df.tail(3)  # 괄호 안 숫자로 행 수 지정

마지막 3명:


,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동
7,P008,한소희,102,여,160,65,145,95,156,금연,가끔
8,P009,윤태완,28,남,178,82,118,78,89,비흡연,규칙적
9,P010,송지원,63,여,155,48,120,92,134,흡연,안함


### 데이터 구조 속성 파악

데이터를 다루기 전에 데이터의 구조를 파악하는 것이 중요합니다.

**주요 속성**
- `shape`: 배열의 형태 (행, 열 개수) - 환자 수와 특성 개수 확인
- `ndim`: 배열의 차원 수 - 벡터(1차원)인지 행렬(2차원)인지 구분  
- `dtype`: 데이터 타입 - 정수형, 실수형, 문자형 등 확인
- `size`: 전체 요소 개수 - 총 데이터 포인트 수 파악


In [ ]:
# 기본 데이터 정보
print(f"전체 환자 수: {df.shape[0]}명")  # 행 개수 = 환자 수
print(f"수집 항목 수: {df.shape[1]}개")  # 열 개수 = 측정 항목 수
print(f"전체 데이터 포인트: {df.size}개")  # 전체 셀 개수

전체 환자 수: 10명
수집 항목 수: 11개
전체 데이터 포인트: 110개


In [ ]:
# 데이터 타입 정보
data_tyes = df.dtypes  # 각 컬럼의 데이터 타입 확인
data_tyes

,0
환자ID,object
이름,object
나이,int64
성별,object
키,int64
체중,int64
수축기혈압,int64
이완기혈압,int64
혈당,int64
흡연,object


In [ ]:
# 상세 데이터 구조
df.info()  # 컬럼별 타입, 결측치 개수, 메모리 사용량 등 종합 정보

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 11 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   환자ID    10 non-null     object
 1   이름      10 non-null     object
 2   나이      10 non-null     int64 
 3   성별      10 non-null     object
 4   키       10 non-null     int64 
 5   체중      10 non-null     int64 
 6   수축기혈압   10 non-null     int64 
 7   이완기혈압   10 non-null     int64 
 8   혈당      10 non-null     int64 
 9   흡연      10 non-null     object
 10  운동      10 non-null     object
dtypes: int64(6), object(5)
memory usage: 1012.0+ bytes


### 🤔 실습 문제

위에서 불러온 **df 데이터**에서 다음을 확인해보세요.
1. **데이터 크기** (행, 열 개수) 출력
2. **처음 3행**과 **마지막 2행** 출력


In [ ]:
# 여기에 작성하세요


In [ ]:
# @title 정답

# 1. 데이터 크기 확인
print("데이터 크기 확인")
print(f"행 개수: {df.shape[0]}")
print(f"열 개수: {df.shape[1]}")
print(f"전체 크기: {df.shape}")

# 2. 처음 3행과 마지막 2행 출력
print("처음 3행:")
print(df.head(3))
print("=" * 100)
print("마지막 2행:")
print(df.tail(2))

데이터 크기 확인
행 개수: 10
열 개수: 11
전체 크기: (10, 11)
처음 3행:
   환자ID   이름  나이 성별    키  체중  수축기혈압  이완기혈압   혈당   흡연   운동
0  P001  김철수  45  남  175  70    120     80   95  비흡연  규칙적
1  P002  이영희  67  여  162  58    140     90  180   금연   가끔
2  P003  박민수  32  남  180  85    110     70   88  비흡연   안함
마지막 2행:
   환자ID   이름  나이 성별    키  체중  수축기혈압  이완기혈압   혈당   흡연   운동
8  P009  윤태완  28  남  178  82    118     78   89  비흡연  규칙적
9  P010  송지원  63  여  155  48    120     92  134   흡연   안함


### 기본 통계 정보 확인

숫자형 데이터의 분포와 특성을 파악하고, 범주형 데이터의 빈도를 확인합니다.


#### describe() 메서드
- **count**: 결측값이 아닌 데이터 개수
- **mean**: 평균값
- **std**: 표준편차 (데이터 분포 정도)
- **min**: 최솟값
- **25%, 50%, 75%**: 1분위수, 중앙값, 3분위수
- **max**: 최댓값


In [ ]:
# 숫자형 데이터 기본 통계 요약
df.describe()  # count, mean, std, min, 25%, 50%, 75%, max 값 출력

,나이,키,체중,수축기혈압,이완기혈압,혈당
count,10.000000,10.000000,10.00000,10.000000,10.000000,10.000000
mean,45.400000,167.500000,66.30000,125.800000,83.500000,114.400000
std,25.764748,8.746428,12.19335,11.390054,7.947746,31.763711
min,12.000000,155.000000,48.00000,110.000000,70.000000,88.000000
25%,29.750000,160.500000,58.50000,118.500000,78.500000,92.750000
50%,38.000000,167.500000,66.50000,122.500000,83.500000,100.000000
75%,58.500000,174.250000,73.75000,133.750000,89.500000,128.000000
max,102.000000,180.000000,85.00000,145.000000,95.000000,180.000000


#### value_counts() 메서드
- 범주형 데이터의 각 값별 빈도수 계산
- 데이터 분포 파악에 유용


In [ ]:
# 범주형 데이터 분포 확인
print("성별 분포:")
df["성별"].value_counts()  # 각 값의 개수 세기

성별 분포:


,count
성별,
남,5
여,5


In [ ]:
print("흡연 상태 분포:")
df["흡연"].value_counts()

흡연 상태 분포:


,count
흡연,
비흡연,5
흡연,3
금연,2


In [ ]:
print("운동 습관 분포:")
df["운동"].value_counts()

운동 습관 분포:


,count
운동,
규칙적,4
가끔,3
안함,3


## Pandas의 인덱싱과 슬라이싱

DataFrame에서 분석 목적에 맞는 특정 부분의 데이터를 선택하고 추출하는 다양한 방법을 다룹니다.

이는 데이터 분석에서 가장 중요하고 자주 사용되는 기능입니다.


### 컬럼 선택

분석하고자 하는 특정 변수(컬럼)만 선택하여 작업하는 방법입니다.

**컬럼 선택 문법**
- **단일 컬럼**: `df['컬럼명']` → Series 반환
- **복수 컬럼**: `df[['컬럼1', '컬럼2']]` → DataFrame 반환

**주의사항**
- 대괄호 1개 vs 2개의 차이점 명확히 구분
- 컬럼명은 정확히 입력 (대소문자, 공백 주의)



In [ ]:
df.head(3)

,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동
0,P001,김철수,45,남,175,70,120,80,95,비흡연,규칙적
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔
2,P003,박민수,32,남,180,85,110,70,88,비흡연,안함


In [ ]:
# 단일 컬럼 선택 → Series 반환
ages = df["나이"]  # 대괄호 한 개 = Series 반환

print(f"데이터 타입: {type(ages)}")

print("처음 5개 값:")
ages.head()

데이터 타입: <class 'pandas.core.series.Series'>
처음 5개 값:


,나이
0,45
1,67
2,32
3,29
4,12


In [ ]:
# 여러 컬럼 선택 → DataFrame 반환
basic_info = df[["이름", "키", "체중"]]  # 대괄호 두 개 = DataFrame 반환

print(f"데이터 타입: {type(basic_info)}")

basic_info.head()

데이터 타입: <class 'pandas.core.frame.DataFrame'>


,이름,키,체중
0,김철수,175,70
1,이영희,162,58
2,박민수,180,85
3,최지은,158,52
4,정도영,170,75


In [ ]:
# 미션: ['이름', '나이', '성별', '키', '체중'] 뽑기

# 여기에 작성하세요


In [ ]:
# 컬럼 순서 변경도 가능
reordered = df[["성별", "이름", "나이"]]  # 원하는 순서로 컬럼배치 가능
print("컬럼 순서 변경:")
reordered.head(3)

컬럼 순서 변경:


,성별,이름,나이
0,남,김철수,45
1,여,이영희,67
2,남,박민수,32


## 고급 데이터 인덱싱

DataFrame에서 특정 데이터를 선택하는 방법에는 2가지가 있습니다.

#### iloc vs loc

| 구분 | iloc | loc |
|------|------|-----|
| **기준** | 위치(숫자) | 라벨(문자열) or 인덱스 번호 |
| **인덱스** | 0, 1, 2... | 0, 1, 2... 또는 'P001', 'P002'... |
| **슬라이싱** | 끝 미포함 (`0:3` = 0,1,2) | 끝 포함 (`'A':'C'` = A,B,C) |
| **사용법** | `df.iloc[0]` | `df.loc['환자A']` |
| **장점** | 위치가 명확 | 의미가 직관적 |
| **단점** | 순서 변경시 위험 | 라벨이 있어야 함 |


### iloc: 위치 기반 인덱싱

**iloc**은 **i**nteger **loc**ation의 줄임말로, 숫자 위치(0부터 시작)를 기반으로 데이터를 선택합니다.


| 특징 | 설명 | 예시 |
|------|------|------|
| **기본 원리** | 위치 기반 (0부터 시작하는 숫자) | `df.iloc[0]` = 첫 번째 행 |
| **슬라이싱 규칙** | 끝 인덱스 미포함 | `df.iloc[0:3]` = 0,1,2번째 (3개) |
| **음수 지원** | 뒤에서부터 접근 | `df.iloc[-1]` = 마지막 행 |
| **리스트 선택** | 불연속 위치를 한번에 선택 | `df.iloc[[0,2,4]]` = 0,2,4번째 |
| **적합한 상황** | 데이터 위치를 정확히 알 때 | 처음 n개, 마지막 n개 선택 시 |

**iloc 사용 시 주의사항**
- 순수 숫자 인덱스만 사용 (컬럼명 사용 불가)
- 데이터 순서 변경 시 결과가 달라질 수 있음
- 슬라이싱 시 끝 번호는 포함되지 않음
- 리스트 형태로 여러 위치 선택 시 대괄호 안에 리스트 형태로 입력

```python
# iloc 문법
# df.iloc[행위치, 열위치] -> 단일 데이터
# df.iloc[행범위, 열범위] -> 다중 데이터
```


#### iloc로 행 선택


In [ ]:
# 1. 단일 행 선택 → Series 반환
print("첫 번째 환자 정보")
first_patient = df.iloc[0]  # 0번째 행 선택
first_patient.head()

첫 번째 환자 정보


,0
환자ID,P001
이름,김철수
나이,45
성별,남
키,175


In [ ]:
# 2. 여러 행 선택 (연속) → DataFrame 반환
print("처음 3명 환자 정보 (연속)")
first_three = df.iloc[0:3]  # 0~2번째 행 선택 (3번째 미포함)
first_three.head()

처음 3명 환자 정보 (연속)


,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동
0,P001,김철수,45,남,175,70,120,80,95,비흡연,규칙적
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔
2,P003,박민수,32,남,180,85,110,70,88,비흡연,안함


In [ ]:
# 3. 여러 행 선택 (불연속) → DataFrame 반환
print("1,3,5번째 환자들 (불연속)")
selected_patients = df.iloc[[1, 3, 5]]  # 1,3,5번째 행 선택
selected_patients.head()

1,3,5번째 환자들 (불연속)


,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔
3,P004,최지은,29,여,158,52,115,75,92,비흡연,규칙적
5,P006,장미래,41,여,165,60,125,82,102,비흡연,규칙적


In [ ]:
# 4. 마지막 행 선택 → Series 반환
print("마지막 환자 정보")
last_patient = df.iloc[-1]  # 마지막 행 선택
last_patient.head()

마지막 환자 정보


,9
환자ID,P010
이름,송지원
나이,63
성별,여
키,155


In [ ]:
# 5. 마지막 3명 선택 → DataFrame 반환
print("마지막 3명 환자")
last_three = df.iloc[-3:]  # 마지막 3개 행 선택
last_three.head()

마지막 3명 환자


,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동
7,P008,한소희,102,여,160,65,145,95,156,금연,가끔
8,P009,윤태완,28,남,178,82,118,78,89,비흡연,규칙적
9,P010,송지원,63,여,155,48,120,92,134,흡연,안함


#### iloc로 열 선택


In [ ]:
# 1. 단일 열 선택 (위치 기반) → Series 반환
print("두 번째 컬럼 선택 (iloc)")
second_column = df.iloc[:, 1]  # 모든 행, 1번째 열 선택
second_column.head()

두 번째 컬럼 선택 (iloc)


,이름
0,김철수
1,이영희
2,박민수
3,최지은
4,정도영


In [ ]:
# 2. 여러 열 선택 (연속) → DataFrame 반환
print("0~3번째 컬럼들 (연속)")
columns_1to3 = df.iloc[:, :4]  # 모든 행, 0~3번째 열 선택
columns_1to3.head()

0~3번째 컬럼들 (연속)


,환자ID,이름,나이,성별
0,P001,김철수,45,남
1,P002,이영희,67,여
2,P003,박민수,32,남
3,P004,최지은,29,여
4,P005,정도영,12,남


In [ ]:
# 3. 여러 열 선택 (불연속) → DataFrame 반환
print("특정 위치의 컬럼들 선택 (불연속)")
selected_columns = df.iloc[:, [1, 3, 5]]  # 모든 행, 1,3,5번째 열 선택
selected_columns.head()

특정 위치의 컬럼들 선택 (불연속)


,이름,성별,체중
0,김철수,남,70
1,이영희,여,58
2,박민수,남,85
3,최지은,여,52
4,정도영,남,75


In [ ]:
# 4. 행과 열 동시 선택 → DataFrame 반환
print("처음 3명의 특정 컬럼들")
subset = df.iloc[0:3, [1, 2, 3]]  # 0~2번째 행, 1,2,3번째 열 선택
subset

처음 3명의 특정 컬럼들


,이름,나이,성별
0,김철수,45,남
1,이영희,67,여
2,박민수,32,남


### loc: 라벨 기반 인덱싱

**loc**은 라벨(인덱스명, 컬럼명)을 기반으로 데이터를 선택합니다.

**Index는 각 행의 번호나 이름을 의미**합니다.



| 특징 | 설명 | 예시 |
|------|------|------|
| **기본 원리** | 라벨 기반 (인덱스명, 컬럼명 사용) | `df.loc['P001']` = P001 환자 |
| **슬라이싱 규칙** | 끝 라벨 포함 | `df.loc['A':'C']` = A,B,C 모두 포함 |
| **조건부 선택** | Boolean 마스크 활용 | `df.loc[df['나이'] > 50]` |
| **리스트 선택** | 불연속 라벨을 한번에 선택 | `df.loc[['P001','P003']]` |
| **적합한 상황** | 의미있는 이름으로 접근할 때 | 특정 환자, 특정 컬럼 선택 시 |

**loc 사용 시 주의사항**
- 인덱스나 컬럼에 해당 라벨이 존재해야 함
- 슬라이싱 시 끝 라벨도 포함됨 (iloc와 다름)
- 조건부 선택 시 Boolean 마스크 결과가 True인 행만 반환
- 숫자 인덱스와 문자 라벨 모두 사용 가능

```python
# loc 문법
# df.iloc[행라벨, 열라벨] -> 단일 데이터
# df.iloc[행범위, 열범위] -> 다중 데이터
# df.loc[조건, 열라벨] -> 조건부 필터링
```


#### Index(인덱스) 이해하기

**Index = 각 행의 번호나 이름**을 나타냅니다. 기본적으로는 숫자(0,1,2...)지만, 의미있는 이름으로 변경할 수 있습니다.


In [ ]:
# 현재 인덱스 확인 (기본: 숫자)
print(f"인덱스 타입: {type(df.index)}")

df.index.tolist()

인덱스 타입: <class 'pandas.core.indexes.range.RangeIndex'>


[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [ ]:
df.head(3)  # 기본형 행번호

,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동
0,P001,김철수,45,남,175,70,120,80,95,비흡연,규칙적
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔
2,P003,박민수,32,남,180,85,110,70,88,비흡연,안함


In [ ]:
# 환자ID를 인덱스로 변경
df_indexed = df.set_index("환자ID")

print(f"새 인덱스 타입: {type(df_indexed.index)}")

print(f"새 인덱스 값: {df_indexed.index.tolist()}")

새 인덱스 타입: <class 'pandas.core.indexes.base.Index'>
새 인덱스 값: ['P001', 'P002', 'P003', 'P004', 'P005', 'P006', 'P007', 'P008', 'P009', 'P010']


In [ ]:
df_indexed.head(3)  # 라벨형 행번호

,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동
환자ID,,,,,,,,,,
P001,김철수,45,남,175,70,120,80,95,비흡연,규칙적
P002,이영희,67,여,162,58,140,90,180,금연,가끔
P003,박민수,32,남,180,85,110,70,88,비흡연,안함


#### loc로 행 선택


In [ ]:
# 환자ID를 인덱스로 설정한 DataFrame 사용
df_indexed = df.set_index("환자ID")

In [ ]:
# 1. 단일 행 선택 (기본 숫자 인덱스) → Series 반환
first_patient = df.loc[0]  # 기본 인덱스 0번째 행 선택
first_patient.head()

,0
환자ID,P001
이름,김철수
나이,45
성별,남
키,175


In [ ]:
# 2. 단일 행 선택 (환자ID 인덱스) → Series 반환
label_petient = df_indexed.loc["P001"]  # 환자ID 'P001' 행 선택
label_petient.head()

,P001
이름,김철수
나이,45
성별,남
키,175
체중,70


In [ ]:
# 3. 여러 행 선택 (환자ID 기준) → DataFrame 반환
multiple_patients = df_indexed.loc[["P001", "P003", "P005"]]  # 여러 환자ID 행 선택
multiple_patients.head()

,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동
환자ID,,,,,,,,,,
P001,김철수,45,남,175,70,120,80,95,비흡연,규칙적
P003,박민수,32,남,180,85,110,70,88,비흡연,안함
P005,정도영,12,남,170,75,135,85,110,흡연,가끔


In [ ]:
# 4. 범위로 선택 (환자ID 기준 - 끝 포함!) → DataFrame 반환
range_patients = df_indexed.loc["P001":"P003"]  # P001~P003 범위의 행 선택
range_patients.head()

,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동
환자ID,,,,,,,,,,
P001,김철수,45,남,175,70,120,80,95,비흡연,규칙적
P002,이영희,67,여,162,58,140,90,180,금연,가끔
P003,박민수,32,남,180,85,110,70,88,비흡연,안함


#### loc로 열 선택


In [ ]:
# 1. 특정 환자의 특정 정보만 → Series 반환
bp_info = df_indexed.loc["P001", ["수축기혈압", "이완기혈압"]]  # P001 행의 혈압 컬럼들 선택
bp_info

,P001
수축기혈압,120
이완기혈압,80


In [ ]:
# 2. 여러 환자의 특정 컬럼들 → DataFrame 반환
basic_info = df_indexed.loc[
    ["P001", "P002"], ["이름", "나이", "성별"]
]  # 여러 행과 기본정보 컬럼들 선택
basic_info

,이름,나이,성별
환자ID,,,
P001,김철수,45,남
P002,이영희,67,여


In [ ]:
# 3. 범위로 행과 열 동시 선택 → DataFrame 반환
range_data = df.loc[0:2, "나이":"키"]  # 0~2번째 행, 나이~키 컬럼 범위 선택
range_data

,나이,성별,키
0,45,남,175
1,67,여,162
2,32,남,180


In [ ]:
# 4. 단일 값 접근 → 스칼라 값 반환
age_value = df_indexed.loc["P001", "나이"]  # P001 행의 나이 컬럼 단일값 선택
print(f"나이: {age_value}세")

나이: 45세


### iloc vs loc 비교 실습


In [ ]:
# 1. 첫 번째 환자 이름 접근
print("iloc 사용:", df.iloc[0, 1])  # 0번째 행, 1번째 열 (이름)
print("loc 사용:", df_indexed.loc["P001", "이름"])  # 0번째 행, '이름' 컬럼

iloc 사용: 김철수
loc 사용: 김철수


In [ ]:
# 2. 처음 3명의 이름과 나이
print("iloc 사용:")
df.iloc[0:3, [1, 2]]  # 0~2행, 1,2열

iloc 사용:


,이름,나이
0,김철수,45
1,이영희,67
2,박민수,32


In [ ]:
print("loc 사용:")
df.loc[0:2, ["이름", "나이"]]  # 0~2행, 특정 컬럼

loc 사용:


,이름,나이
0,김철수,45
1,이영희,67
2,박민수,32


In [ ]:
# 3. 마지막 환자 이름
print("iloc 사용:", df.iloc[-1, 1])  # 마지막 행, 1번째 열
print("loc 사용:", df.loc[df.index[-1], "이름"])  # 마지막 인덱스 값 사용

iloc 사용: 송지원
loc 사용: 송지원


### 조건부 필터링

데이터 분석에서 특정 조건을 만족하는 데이터만 골라내는 것은 매우 중요한 기능입니다.

pandas에서는 조건문을 사용해 원하는 데이터를 쉽게 필터링할 수 있습니다.

**Boolean 마스크 작동 원리**
- **Boolean 마스크**: 조건 결과로 True/False 배열 생성
  
  `df['나이'] >= 50` → `[False, True, True, False, True]`
- **마스크 적용**: True인 위치의 데이터만 추출
  
  `df[df['나이'] >= 50]` → 조건 만족하는 모든 환자 정보

In [ ]:
# 단일 조건 - Boolean 마스크 단계별 확인
hypertension_mask = df["수축기혈압"] >= 140

print(f"고혈압 마스크: {hypertension_mask.values}")
print("=" * 30)

# 마스크 적용하여 고혈압 환자 선택
hypertension_patients = df[hypertension_mask]
print(f"고혈압 환자 수: {len(hypertension_patients)}명")

hypertension_patients

고혈압 마스크: [False  True False False False False False  True False False]
고혈압 환자 수: 2명


,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔
7,P008,한소희,102,여,160,65,145,95,156,금연,가끔


#### 복합 조건 (여러 조건 조합)

실제 데이터 분석에서는 여러 조건을 동시에 만족하는 데이터를 찾아야 하는 경우가 많습니다.

**복합 조건 연산자**
- `&` (AND): 모든 조건이 True일 때만 True
- `|` (OR): 하나 이상의 조건이 True이면 True
- `~` (NOT): 조건 결과를 반대로 변환

**주의사항**
- 각 조건을 괄호 `()`로 묶어야 함
- 괄호 없이 쓰면 pandas가 의도와 다르게 해석함
- `and`, `or` 대신 `&`, `|` 사용


In [ ]:
# 복합 조건 - 여러 조건 조합
age_mask = df["나이"] >= 50
bp_mask = df["수축기혈압"] >= 140

In [ ]:
# AND 조건: 두 조건을 모두 만족하는 환자

senior_hypertension_patients = df[age_mask & bp_mask]

print(f"50세 이상 + 고혈압 환자수: {len(senior_hypertension_patients)}명")

senior_hypertension_patients

50세 이상 + 고혈압 환자수: 2명


,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔
7,P008,한소희,102,여,160,65,145,95,156,금연,가끔


In [ ]:
# OR 조건: 하나라도 만족하는 환자
high_risk_patients = df[age_mask | bp_mask]

print(f"50세 이상 또는 고혈압 환자수: {len(high_risk_patients)}명")

high_risk_patients

50세 이상 또는 고혈압 환자수: 3명


,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔
7,P008,한소희,102,여,160,65,145,95,156,금연,가끔
9,P010,송지원,63,여,155,48,120,92,134,흡연,안함


In [ ]:
# NOT 조건: 조건을 만족하지 않는 환자
young_normal_patients = df[~age_mask & ~bp_mask]

print(f"50세 미만 + 정상혈압 환자수: {len(young_normal_patients)}명")

young_normal_patients

50세 미만 + 정상혈압 환자수: 7명


,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동
0,P001,김철수,45,남,175,70,120,80,95,비흡연,규칙적
2,P003,박민수,32,남,180,85,110,70,88,비흡연,안함
3,P004,최지은,29,여,158,52,115,75,92,비흡연,규칙적
4,P005,정도영,12,남,170,75,135,85,110,흡연,가끔
5,P006,장미래,41,여,165,60,125,82,102,비흡연,규칙적
6,P007,오성진,35,남,172,68,130,88,98,흡연,안함
8,P009,윤태완,28,남,178,82,118,78,89,비흡연,규칙적


In [ ]:
# 3. 조건부로 파생컬럼 추가
df_copy = df.copy()  # 원본 데이터프레임 복사

# 혈압상태 컬럼 추가하여 데이터 적재
df_copy.loc[df_copy["수축기혈압"] >= 140, "혈압상태"] = "고혈압"
df_copy.loc[df_copy["수축기혈압"] < 140, "혈압상태"] = "정상"

df_copy.head()

,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동,혈압상태
0,P001,김철수,45,남,175,70,120,80,95,비흡연,규칙적,정상
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔,고혈압
2,P003,박민수,32,남,180,85,110,70,88,비흡연,안함,정상
3,P004,최지은,29,여,158,52,115,75,92,비흡연,규칙적,정상
4,P005,정도영,12,남,170,75,135,85,110,흡연,가끔,정상


In [ ]:
# 혈압 상태 분포 확인
bp_status_counts = df_copy["혈압상태"].value_counts()
print("혈압 상태 분포:")
bp_status_counts

혈압 상태 분포:


,count
혈압상태,
정상,8
고혈압,2


### 🤔 실습 문제


위에서 불러온 df 데이터에서 다음을 확인해보세요.

1. **iloc 사용**: 5~7번째 환자의 이름, 나이, 성별 정보 출력
2. **loc 사용**: 'P003' 환자의 수축기혈압, 이완기혈압, 혈당 정보 출력  
3. **조건부 필터링**: 여성 환자 중 키가 160cm 이상인 환자들의 전체 정보 출력


In [ ]:
# 여기에 작성하세요


In [ ]:
# @title 정답

# 1. iloc 사용 - 5~7번째 환자의 기본 정보
print("1. iloc으로 5~7번째 환자 정보:")
patients_5to7 = df.iloc[4:7, [1, 2, 3]]  # 4~6번째 인덱스, 이름/나이/성별 컬럼
print(patients_5to7)
print()

# 2. loc 사용 - P003 환자의 건강 정보
print("2. loc으로 P003 환자의 건강 정보:")
df_indexed = df.set_index("환자ID")
p003_health = df_indexed.loc["P003", ["수축기혈압", "이완기혈압", "혈당"]]
print(p003_health)
print()

# 3. 조건부 필터링 - 여성 + 키 160 이상
print("3. 조건부 필터링: 여성 환자 중 키 160cm 이상:")
tall_women = df[(df["성별"] == "여") & (df["키"] >= 160)]
print(f"조건을 만족하는 환자 수: {len(tall_women)}명")
print(tall_women)

1. iloc으로 5~7번째 환자 정보:
    이름  나이 성별
4  정도영  12  남
5  장미래  41  여
6  오성진  35  남

2. loc으로 P003 환자의 건강 정보:
수축기혈압    110
이완기혈압     70
혈당        88
Name: P003, dtype: object

3. 조건부 필터링: 여성 환자 중 키 160cm 이상:
조건을 만족하는 환자 수: 3명
   환자ID   이름   나이 성별    키  체중  수축기혈압  이완기혈압   혈당   흡연   운동
1  P002  이영희   67  여  162  58    140     90  180   금연   가끔
5  P006  장미래   41  여  165  60    125     82  102  비흡연  규칙적
7  P008  한소희  102  여  160  65    145     95  156   금연   가끔


# DataFrame 데이터 조작

DataFrame에서 데이터를 추가, 수정, 삭제하는 다양한 방법을 체계적으로 알아봅니다.

기본적인 데이터 조작부터 고급 인덱싱, 데이터 삭제까지 단계별로 다룹니다.

## 1. 기본 데이터 조작 (추가/수정)

가장 기본적인 데이터 조작 방법으로,  
컬럼(속성·특징)과 행(개체·관측 대상)을 추가·수정하고 관리하며 데이터 구조를 확장·정비하는 방법을 다룹니다.

### 컬럼 추가/수정

#### 컬럼 추가

기존 DataFrame에 새로운 속성(특징)을 추가하여 데이터의 표현 범위를 확장하는 방법을 다룹니다.  
주로 계산된 지표, 분류 결과, 분석용 파생 변수를 생성할 때 사용됩니다.


**기본 컬럼 생성 문법**
- **단일 값 할당**: `df['새컬럼'] = 값`
- **계산 결과 할당**: `df['새컬럼'] = df['기존컬럼1'] + df['기존컬럼2']`
- **조건부 할당**: `df.loc[조건, '새컬럼'] = 값`


In [ ]:
# 기존 데이터
df.head(3)

,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동
0,P001,김철수,45,남,175,70,120,80,95,비흡연,규칙적
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔
2,P003,박민수,32,남,180,85,110,70,88,비흡연,안함


##### 단일 컬럼 추가

하나의 새로운 컬럼을 생성하고, 모든 행에 대해 동일한 단일 값을 일괄로 부여하는 방식입니다.

In [ ]:
# 단일 값으로 새 컬럼 추가
df["병원"] = "서울대병원"

df.head(3)

,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동,병원
0,P001,김철수,45,남,175,70,120,80,95,비흡연,규칙적,서울대병원
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔,서울대병원
2,P003,박민수,32,남,180,85,110,70,88,비흡연,안함,서울대병원


##### 다중 컬럼 추가

여러 개의 새로운 컬럼을 한 번에 생성하고, 각 컬럼에 대해 단일 값을 동시에 부여하는 방식입니다.  
분석 전 기본 분류 기준이나 초기 상태값을 설정할 때 유용합니다.

In [ ]:
# 분석용 기본 상태 컬럼들 동시 추가
df[["분석대상", "데이터출처"]] = [True, "내원환자"]

df.head(3)

,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동,병원,분석대상,데이터출처
0,P001,김철수,45,남,175,70,120,80,95,비흡연,규칙적,서울대병원,True,내원환자
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔,서울대병원,True,내원환자
2,P003,박민수,32,남,180,85,110,70,88,비흡연,안함,서울대병원,True,내원환자


##### 계산 결과 컬럼 추가

**BMI 컬럼 추가**

환자의 체중과 키 정보를 활용하여 BMI 값을 계산한 새로운 컬럼을 추가합니다.

BMI(Body Mass Index, 체질량지수)는 신장과 체중을 이용하여 비만도를 나타내는 지수로, 건강 상태를 평가하는 중요한 지표입니다.

**BMI 계산 공식**
```
BMI = 체중(kg) ÷ (키(m))²
```


In [ ]:
# 계산으로 새 컬럼 추가 - BMI 계산
df["BMI"] = df["체중"] / ((df["키"] / 100) ** 2)  # 계산 결과: Series타입

df["BMI"] = df["BMI"].round(1)  # 소수점 첫째자리까지 표시하여 반올림

df.head(3)

,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동,병원,분석대상,데이터출처,BMI
0,P001,김철수,45,남,175,70,120,80,95,비흡연,규칙적,서울대병원,True,내원환자,22.9
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔,서울대병원,True,내원환자,22.1
2,P003,박민수,32,남,180,85,110,70,88,비흡연,안함,서울대병원,True,내원환자,26.2


##### 조건부 컬럼 생성
**BMI 범주 분류 컬럼 추가**

계산된 BMI 값을 의미있는 범주로 분류하는 새로운 컬럼을 추가합니다.

**BMI 범주 분류 기준**
- 저체중: 18.5 미만
- 정상체중: 18.5~22.9
- 과체중: 23.0~24.9
- 경도비만: 25.0~29.9
- 고도비만: 30.0 이상

In [ ]:
# BMI 기준으로 비만도 분류 (조건부 할당 방식)
# 조건에 부합하지 않는 경우 NaN(데이터 없음, 결측치) 할당

df.loc[df["BMI"] < 18.5, "BMI_범주"] = "저체중"
df.loc[(df["BMI"] >= 18.5) & (df["BMI"] < 23), "BMI_범주"] = "정상체중"
# df.loc[(df['BMI'] >= 23) & (df['BMI'] < 25), 'BMI_범주'] = '과체중'
df.loc[(df["BMI"] >= 25) & (df["BMI"] < 30), "BMI_범주"] = "경도비만"
df.loc[df["BMI"] >= 30, "BMI_범주"] = "고도비만"

print(df["BMI_범주"].value_counts())

print("=" * 30)

print("추가된 DataFrame")
df.head()

BMI_범주
정상체중    5
경도비만    4
Name: count, dtype: int64
추가된 DataFrame


,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동,병원,분석대상,데이터출처,BMI,BMI_범주
0,P001,김철수,45,남,175,70,120,80,95,비흡연,규칙적,서울대병원,True,내원환자,22.9,정상체중
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔,서울대병원,True,내원환자,22.1,정상체중
2,P003,박민수,32,남,180,85,110,70,88,비흡연,안함,서울대병원,True,내원환자,26.2,경도비만
3,P004,최지은,29,여,158,52,115,75,92,비흡연,규칙적,서울대병원,True,내원환자,20.8,정상체중
4,P005,정도영,12,남,170,75,135,85,110,흡연,가끔,서울대병원,True,내원환자,26.0,경도비만


In [ ]:
# 조건 결과 계산하여 추가
df["고위험군"] = df["수축기혈압"] >= 140  # bool값 추가

df.head(3)

,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동,병원,분석대상,데이터출처,BMI,BMI_범주,고위험군
0,P001,김철수,45,남,175,70,120,80,95,비흡연,규칙적,서울대병원,True,내원환자,22.9,정상체중,False
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔,서울대병원,True,내원환자,22.1,정상체중,True
2,P003,박민수,32,남,180,85,110,70,88,비흡연,안함,서울대병원,True,내원환자,26.2,경도비만,False


#### 컬럼 수정

이미 존재하는 컬럼의 값을 새로운 기준이나 조건에 맞게 수정하는 방법을 다룹니다.  
주로 데이터 정정, 기준 변경, 분석 단계에 따른 재분류가 필요할 때 사용됩니다.

##### 일괄 수정

모든 행의 특정 컬럼을 동일한 값으로 변경하는 방법입니다.

In [ ]:
# 전체 컬럼 값 일괄 수정
df["병원"] = "을지대학교병원"

df.head(3)

,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동,병원,분석대상,데이터출처,BMI,BMI_범주,고위험군
0,P001,김철수,45,남,175,70,120,80,95,비흡연,규칙적,을지대학교병원,True,내원환자,22.9,정상체중,False
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔,을지대학교병원,True,내원환자,22.1,정상체중,True
2,P003,박민수,32,남,180,85,110,70,88,비흡연,안함,을지대학교병원,True,내원환자,26.2,경도비만,False


##### ⭐️ 조건부 수정

특정 조건을 만족하는 행의 컬럼값을 선별적으로 수정하는 방법입니다.  
실무 데이터 분석과 전처리에서 압도적으로 많이 사용되는 수정 방식입니다.

- 상태·라벨·등급 컬럼을 갱신할 때 사용
(예: 위험군, 관리 단계, 진료 분류)
- 모델 학습을 위한 전처리 단계에서
(예: 기준 초과 여부, 이상치 플래그)

📌 핵심 포인트
* ```df.loc[조건, 컬럼] = 값```
* 조건은 Boolean mask
* 값은 스칼라, 리스트, Series 모두 가능

**조건부 수정 (단일 컬럼)**

특정 조건을 만족하는 행의 데이터만 선별적으로 수정하는 방법입니다.

In [ ]:
# '병원' 컬럼 값을 조건에 따라 수정
df.loc[df["나이"] >= 50, "병원"] = "서울대학교병원 노인의학과"
df.loc[df["나이"] < 50, "병원"] = "서울대학교병원 가정의학과"

df[["이름", "나이", "병원"]].head(3)

,이름,나이,병원
0,김철수,45,서울대학교병원 가정의학과
1,이영희,67,서울대학교병원 노인의학과
2,박민수,32,서울대학교병원 가정의학과


**조건부 수정 (다중 컬럼)**

특정 조건을 기준으로 여러 컬럼의 값을 한 번에 수정하는 방식입니다.

```['고위험군', '병원'] 컬럼 값을 조건에 따라 수정```

In [ ]:
# 환자 조건에 따라 분석 대상 환자 분류
# - 혈당 126 이상 또는
# - 수축기혈압 140 이상

analysis_target_mask = (df["혈당"] >= 126) | (df["수축기혈압"] >= 140)

# '분석대상', '데이터출처' 컬럼값을 조건에 따라 수정
df.loc[analysis_target_mask, ["분석대상", "데이터출처"]] = [True, "내원환자"]
df.loc[~analysis_target_mask, ["분석대상", "데이터출처"]] = [False, "외래환자"]

df[["이름", "혈당", "수축기혈압", "분석대상", "데이터출처"]].head(10)

,이름,혈당,수축기혈압,분석대상,데이터출처
0,김철수,95,120,False,외래환자
1,이영희,180,140,True,내원환자
2,박민수,88,110,False,외래환자
3,최지은,92,115,False,외래환자
4,정도영,110,135,False,외래환자
5,장미래,102,125,False,외래환자
6,오성진,98,130,False,외래환자
7,한소희,156,145,True,내원환자
8,윤태완,89,118,False,외래환자
9,송지원,134,120,True,내원환자


### 🤔 실습 문제

**조건부 컬럼 추가 연습**

현재 DataFrame의 환자 데이터를 분석하여 새로운 분류 컬럼들을 추가해보세요.

1. **'고령환자' 컬럼**: 60세 이상이면 `True`, 그 외는 `False` 할당
2. **'위험도등급' 컬럼**: 고혈압(수축기 140 이상)이면서 BMI 25 이상이면 `"고위험"`, 그 외는 `"일반"` 할당


In [ ]:
# 여기에 작성하세요


In [ ]:
# @title 정답

# 1. 고령환자 컬럼 추가 - 60세 이상 여부
df.loc[df["나이"] >= 60, "고령환자"] = True
df.loc[df["나이"] < 60, "고령환자"] = False

print("고령환자 컬럼 추가 완료")

old_total_count = df["고령환자"].sum()
print(f"고령환자: {old_total_count}명")
print(f"일반환자: {len(df) - old_total_count}명")

print("=" * 60)

# 2. 위험도등급 컬럼 추가 - 고혈압 + 비만 복합 조건
high_risk_condition = (df["수축기혈압"] >= 140) & (df["BMI"] >= 25)
df.loc[high_risk_condition, "위험도등급"] = "고위험"
df.loc[~high_risk_condition, "위험도등급"] = "일반"

print("위험도등급 컬럼 추가 완료")

print("위험도등급 분포:")
print(df["위험도등급"].value_counts())

print("=" * 60)

# 최종 결과 확인
print(df[["이름", "나이", "수축기혈압", "BMI", "고령환자", "위험도등급"]])

고령환자 컬럼 추가 완료
고령환자: 3명
일반환자: 7명
위험도등급 컬럼 추가 완료
위험도등급 분포:
위험도등급
일반     9
고위험    1
Name: count, dtype: int64
    이름   나이  수축기혈압   BMI   고령환자 위험도등급
0  김철수   45    120  22.9  False    일반
1  이영희   67    140  22.1   True    일반
2  박민수   32    110  26.2  False    일반
3  최지은   29    115  20.8  False    일반
4  정도영   12    135  26.0  False    일반
5  장미래   41    125  22.0  False    일반
6  오성진   35    130  23.0  False    일반
7  한소희  102    145  25.4   True   고위험
8  윤태완   28    118  25.9  False    일반
9  송지원   63    120  20.0   True    일반


#### 행 추가

기존 DataFrame에 새로운 행을 추가하여 데이터를 누적하는 방법을 다룹니다.  
주로 새 환자, 신규 기록, 추가 수집된 데이터를 반영할 때 사용됩니다.

##### 단일 행 추가

한 명의 환자 정보만 추가할 때는 `df.loc[]` 방식이 간단하고 직관적입니다.


In [ ]:
print(f"환자 추가 전 총 인원: {len(df)}명")
print(f"환자 데이터프레임 컬럼 종류: {df.columns.tolist()}")  # 컬럼 데이터프레임을 리스트로 변환

환자 추가 전 총 인원: 10명
환자 데이터프레임 컬럼 종류: ['환자ID', '이름', '나이', '성별', '키', '체중', '수축기혈압', '이완기혈압', '혈당', '흡연', '운동', '병원', '분석대상', '데이터출처', 'BMI', 'BMI_범주', '고위험군', '고령환자', '위험도등급']


In [ ]:
# 방법 1: 인덱스 번호로 추가
# 컬럼수와 순서에 맞춰서 리스트 작성해야 함

df.loc[len(df)] = [
    "P011",
    "강동원",
    42,
    "남",
    178,
    73,
    130,
    85,
    98,
    "비흡연",
    "규칙적",
    "서울대학교병원 가정의학과",
    False,
    "외래환자",
    23.0,
    False,
    "정상체중",
    False,
    "일반",
]


print(f"환자 추가 후 총 인원: {len(df)}명")
print("=" * 30)

df.tail(3)

환자 추가 후 총 인원: 11명


,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동,병원,분석대상,데이터출처,BMI,BMI_범주,고위험군,고령환자,위험도등급
8,P009,윤태완,28,남,178,82,118,78,89,비흡연,규칙적,서울대학교병원 가정의학과,False,외래환자,25.9,경도비만,False,False,일반
9,P010,송지원,63,여,155,48,120,92,134,흡연,안함,서울대학교병원 노인의학과,True,내원환자,20.0,정상체중,False,True,일반
10,P011,강동원,42,남,178,73,130,85,98,비흡연,규칙적,서울대학교병원 가정의학과,False,외래환자,23.0,False,정상체중,False,일반


In [ ]:
# 방법 2: 특정 인덱스명으로 추가
# 추가 데이터에서 '환자ID'값 제외

df_indexed.loc["P011"] = ["강동원", 42, "남", 178, 73, 130, 85, 98, "비흡연", "규칙적"]

print(f"환자 추가 후: {len(df_indexed)}명")
print("=" * 30)

df_indexed.tail(3)

환자 추가 후: 11명


,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동
환자ID,,,,,,,,,,
P009,윤태완,28,남,178,82,118,78,89,비흡연,규칙적
P010,송지원,63,여,155,48,120,92,134,흡연,안함
P011,강동원,42,남,178,73,130,85,98,비흡연,규칙적


##### 여러 행 추가

여러 명의 환자 정보를 동시에 추가할 때는 `concat()` 함수를 사용합니다.

**concat() 함수**  
여러 DataFrame을 연결(concatenate)하는 pandas의 핵심 함수입니다.
- `axis=0` (기본값): 행 방향 연결
- `axis=1`: 열 방향 연결

**ignore_index 옵션**
`ignore_index` 옵션은 DataFrame을 연결할 때 인덱스 처리 방식을 결정하는 파라미터입니다.   
데이터를 추가할 때 인덱스 중복 문제를 해결하고 깔끔한 연속 인덱스를 생성하기 위해 사용됩니다.

- `ignore_index=False` (기본값): 추가되는 데이터의 기존 인덱스를 유지하여 인덱스 중복 발생 가능
- `ignore_index=True` (권장): 추가되는 데이터의 기존 인덱스를 무시하고 새로운 연속 인덱스(0,1,2...) 생성

**권장 이유**
- 인덱스 중복으로 인한 데이터 접근 오류 방지
- 일관된 연속 인덱스로 데이터 관리 용이성 향상
- 추후 데이터 분석 시 예측 가능한 인덱스 구조 제공




In [ ]:
# 여러 환자 동시 추가
new_patients_data = [
    [
        "P012",
        "박보검",
        31,
        "남",
        182,
        68,
        118,
        78,
        92,
        "비흡연",
        "규칙적",
        "서울대학교병원 가정의학과",
        False,
        "외래환자",
        20.5,
        False,
        "정상체중",
        False,
        "일반",
    ],
    [
        "P013",
        "김혜수",
        54,
        "여",
        165,
        58,
        145,
        92,
        105,
        "금연",
        "가끔",
        "서울대학교병원 노인의학과",
        True,
        "내원환자",
        21.3,
        True,
        "정상체중",
        False,
        "일반",
    ],
    [
        "P014",
        "이정재",
        63,
        "남",
        175,
        78,
        155,
        98,
        125,
        "흡연",
        "안함",
        "서울대학교병원 노인의학과",
        True,
        "내원환자",
        25.5,
        True,
        "경도비만",
        True,
        "일반",
    ],
    [
        "P015",
        "전지현",
        29,
        "여",
        168,
        52,
        110,
        70,
        88,
        "비흡연",
        "규칙적",
        "서울대학교병원 가정의학과",
        False,
        "외래환자",
        18.4,
        False,
        "저체중",
        False,
        "일반",
    ],
]

# 새로운 데이터프레임 생성
new_df = pd.DataFrame(new_patients_data, columns=df.columns)

# 기준 데이터프레임에 합치기
df_extended = pd.concat([df, new_df], ignore_index=True)

print(f"기존 환자 수: {len(df)}명")
print(f"여러 환자 추가 후: {len(df_extended)}명")

print("=" * 30)

df_extended.tail()

기존 환자 수: 11명
여러 환자 추가 후: 15명


,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동,병원,분석대상,데이터출처,BMI,BMI_범주,고위험군,고령환자,위험도등급
10,P011,강동원,42,남,178,73,130,85,98,비흡연,규칙적,서울대학교병원 가정의학과,False,외래환자,23.0,False,정상체중,False,일반
11,P012,박보검,31,남,182,68,118,78,92,비흡연,규칙적,서울대학교병원 가정의학과,False,외래환자,20.5,False,정상체중,False,일반
12,P013,김혜수,54,여,165,58,145,92,105,금연,가끔,서울대학교병원 노인의학과,True,내원환자,21.3,True,정상체중,False,일반
13,P014,이정재,63,남,175,78,155,98,125,흡연,안함,서울대학교병원 노인의학과,True,내원환자,25.5,True,경도비만,True,일반
14,P015,전지현,29,여,168,52,110,70,88,비흡연,규칙적,서울대학교병원 가정의학과,False,외래환자,18.4,False,저체중,False,일반


#### 단일 값 수정

기존 DataFrame에서 **특정 행의 값(스칼라)을 새로운 정보로 수정**하는 방법을 다룹니다.  
주로 **데이터 오류 수정, 상태 변화 반영, 조건에 따른 결과 업데이트**가 필요할 때 사용됩니다.



In [ ]:
# 첫 번째 환자의 체중 정보 수정 (측정 오류 정정)
df.loc[0, "체중"] = 105  # Series타입

df[["이름", "체중"]].head()

,이름,체중
0,김철수,105
1,이영희,58
2,박민수,85
3,최지은,52
4,정도영,75


## 3. 데이터 삭제

분석에 필요하지 않은 행이나 컬럼을 제거해 **분석 대상만 남기기 위한 데이터 정제 작업**입니다.

### drop() 메서드

**drop() 메서드**
DataFrame에서 불필요한 **행 또는 컬럼을 제거**할 때 사용하는 메서드입니다.  
전처리 과정에서 데이터 구조를 단순화하거나 쓰레기값을 제거할 때 활용합니다.

### 주요 옵션

**inplace 옵션**
원본 DataFrame을 직접 수정할지 여부를 결정합니다.

* `inplace=False` (기본값): 새로운 DataFrame 반환, 원본 유지
* `inplace=True`: 원본 DataFrame 직접 수정하고 `None` 반환

**axis 옵션**
데이터 삭제 방향을 지정하는 옵션으로, DataFrame의 행과 열 중 어느 방향으로 삭제할지를 결정합니다.

* `axis=0` (기본값): 행(인덱스) 삭제
* `axis=1`: 열(컬럼) 삭제


In [ ]:
# 1. 원본 DataFrame 보호 방식 (inplace=False, 기본값)
df_without_hospital = df.drop("병원", axis=1)  # 새로운 DataFrame 반환

print(f"원본 컬럼 수: {len(df.columns)}")
print(f"복사본 컬럼 수: {len(df_without_hospital.columns)}")

원본 컬럼 수: 19
복사본 컬럼 수: 18


### inplace (원본 보호 vs 원본 직접 수정)


In [ ]:
# 2. 원본 DataFrame 직접 수정 방식 (inplace=True)
df_copy = df.copy()  # 복사본 생성
print(f"삭제 전 컬럼 수: {len(df_copy.columns)}")

df_copy.drop("병원", axis=1, inplace=True)  # 원본 직접 수정
print(f"삭제 후 컬럼 수: {len(df_copy.columns)}")

삭제 전 컬럼 수: 19
삭제 후 컬럼 수: 18


### 컬럼 삭제

분석이나 모델에 사용하지 않는 컬럼을 제거할 때 사용합니다.


In [ ]:
# 1. 단일 컬럼 삭제 (새 DataFrame 반환)
df_without_hospital = df.drop("병원", axis=1)

df_without_hospital.head()

,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동,분석대상,데이터출처,BMI,BMI_범주,고위험군,고령환자,위험도등급
0,P001,김철수,45,남,175,105,120,80,95,비흡연,규칙적,False,외래환자,22.9,정상체중,False,False,일반
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔,True,내원환자,22.1,정상체중,True,True,일반
2,P003,박민수,32,남,180,85,110,70,88,비흡연,안함,False,외래환자,26.2,경도비만,False,False,일반
3,P004,최지은,29,여,158,52,115,75,92,비흡연,규칙적,False,외래환자,20.8,정상체중,False,False,일반
4,P005,정도영,12,남,170,75,135,85,110,흡연,가끔,False,외래환자,26.0,경도비만,False,False,일반


In [ ]:
# 2. 여러 컬럼 동시 삭제
df_essential = df.drop(["고령환자", "고위험군"], axis=1)

df_essential.head()

,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동,병원,분석대상,데이터출처,BMI,BMI_범주,위험도등급
0,P001,김철수,45,남,175,105,120,80,95,비흡연,규칙적,서울대학교병원 가정의학과,False,외래환자,22.9,정상체중,일반
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔,서울대학교병원 노인의학과,True,내원환자,22.1,정상체중,일반
2,P003,박민수,32,남,180,85,110,70,88,비흡연,안함,서울대학교병원 가정의학과,False,외래환자,26.2,경도비만,일반
3,P004,최지은,29,여,158,52,115,75,92,비흡연,규칙적,서울대학교병원 가정의학과,False,외래환자,20.8,정상체중,일반
4,P005,정도영,12,남,170,75,135,85,110,흡연,가끔,서울대학교병원 가정의학과,False,외래환자,26.0,경도비만,일반


In [ ]:
# 3. 컬럼 위치(번호) 기준 삭제 - 마지막 컬럼 삭제
df_no_last_col = df.drop(columns=df.columns[-1])
df_no_last_col.head()

,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,흡연,운동,병원,분석대상,데이터출처,BMI,BMI_범주,고위험군,고령환자
0,P001,김철수,45,남,175,105,120,80,95,비흡연,규칙적,서울대학교병원 가정의학과,False,외래환자,22.9,정상체중,False,False
1,P002,이영희,67,여,162,58,140,90,180,금연,가끔,서울대학교병원 노인의학과,True,내원환자,22.1,정상체중,True,True
2,P003,박민수,32,남,180,85,110,70,88,비흡연,안함,서울대학교병원 가정의학과,False,외래환자,26.2,경도비만,False,False
3,P004,최지은,29,여,158,52,115,75,92,비흡연,규칙적,서울대학교병원 가정의학과,False,외래환자,20.8,정상체중,False,False
4,P005,정도영,12,남,170,75,135,85,110,흡연,가끔,서울대학교병원 가정의학과,False,외래환자,26.0,경도비만,False,False


**존재하하지 않는 컬럼 삭제 시 에러발생**

In [ ]:
# 존재하지 않는 컬럼 삭제
df_essential.drop("고령환자", axis=1)  # Error!

df_essential

KeyError: "['고령환자'] not found in axis"

### 행 삭제

분석 대상에서 제외할 데이터가 있을 때 사용합니다.


In [ ]:
print(f"기존 DataFrme 환자수: {len(df)}명")

기존 DataFrme 환자수: 11명


In [ ]:
# 1. 인덱스로 단일 행 삭제 (원본 영향 x)
df_no_first = df.drop(0)  # 삭제된
print(f"첫 번째 환자 제외: {len(df_no_first)}명")

df_no_first[["이름", "나이", "고위험군"]].head()

첫 번째 환자 제외: 10명


,이름,나이,고위험군
1,이영희,67,True
2,박민수,32,False
3,최지은,29,False
4,정도영,12,False
5,장미래,41,False


In [ ]:
# 2. 마지막 행 삭제 (원본 영향 x)
df_no_last = df.drop(df.index[-1])
print(f"마지막 환자 제외: {len(df_no_last)}명")

df_no_last[["이름", "나이", "고위험군"]].tail()

마지막 환자 제외: 10명


,이름,나이,고위험군
5,장미래,41,False
6,오성진,35,False
7,한소희,102,True
8,윤태완,28,False
9,송지원,63,False


In [ ]:
# 3. 여러 행 동시 삭제 (원본 영향 x)
df_selected = df.drop([0, 1, 2])
print(f"처음 3명 제외 후: {len(df_selected)}명")

df_selected[["이름", "나이", "고위험군"]].head()

처음 3명 제외 후: 8명


,이름,나이,고위험군
3,최지은,29,False
4,정도영,12,False
5,장미래,41,False
6,오성진,35,False
7,한소희,102,True


### 조건부 행 삭제

데이터 품질 문제나 분석 제외 기준을 반영할 때 사용합니다.


In [ ]:
df.loc[:, "나이"]

,나이
0,45
1,67
2,32
3,29
4,12
5,41
6,35
7,102
8,28
9,63


In [ ]:
# 분석 제외 조건에 해당하는 환자 제거
# 제외 조건: 나이 20세 미만 또는 100세 초과
exclude_age_mask = (df["나이"] < 20) | (df["나이"] > 100)

df[exclude_age_mask][["환자ID", "이름", "나이", "고위험군"]]

,환자ID,이름,나이,고위험군
4,P005,정도영,12,False
7,P008,한소희,102,True


In [ ]:
# 조건에 부합하는 행들의 인덱스 리스트 반환
df[exclude_age_mask].index

Index([4, 7], dtype='int64')

In [ ]:
# index를 활용하여 행 삭제
df_clean = df.drop(df[exclude_age_mask].index)

print(f"전체 대상 환자 수: {len(df)}명")
print(f"분석 대상 환자 수: {len(df_clean)}명")

전체 대상 환자 수: 11명
분석 대상 환자 수: 9명


### 인덱스 기반 삭제

의미 있는 식별자를 인덱스로 사용하는 경우, 라벨 기준으로 삭제할 수 있습니다.


In [ ]:
# 환자ID를 인덱스로 설정
df_indexed = df.set_index("환자ID")

df_indexed[["이름", "나이", "고위험군"]]

,이름,나이,고위험군
환자ID,,,
P001,김철수,45,False
P002,이영희,67,True
P003,박민수,32,False
P004,최지은,29,False
P005,정도영,12,False
P006,장미래,41,False
P007,오성진,35,False
P008,한소희,102,True
P009,윤태완,28,False


In [ ]:
# 특정 환자 삭제
df_indexed = df_indexed.drop("P001")
print(f"P001 환자 제외 후: {len(df_indexed)}명")

df_indexed[["이름", "나이", "고위험군"]].head(3)

P001 환자 제외 후: 10명


,이름,나이,고위험군
환자ID,,,
P002,이영희,67,True
P003,박민수,32,False
P004,최지은,29,False


In [ ]:
# 여러 환자 동시 삭제
df_indexed = df_indexed.drop(["P002", "P003"])
print(f"P002, P003 제외 후: {len(df_indexed)}명")

df_indexed[["이름", "나이", "고위험군"]].head()

P002, P003 제외 후: 8명


,이름,나이,고위험군
환자ID,,,
P004,최지은,29,False
P005,정도영,12,False
P006,장미래,41,False
P007,오성진,35,False
P008,한소희,102,True


### 행 + 컬럼 동시 삭제

`drop()`은 **행과 컬럼을 동시에 삭제**하는 것도 가능합니다.  
- 행은 인덱스 기준
- 컬럼은 컬럼명 기준


In [ ]:
# 2번째 환자(인덱스 기준) 삭제 + 불필요 컬럼 2개 동시 삭제
df_removed = df.drop(index=df.index[-1], columns=["흡연", "운동"])

print(f"마지막 환자 제외 후: {len(df_removed)}명")

df_removed.tail(3)

마지막 환자 제외 후: 10명


,환자ID,이름,나이,성별,키,체중,수축기혈압,이완기혈압,혈당,병원,분석대상,데이터출처,BMI,BMI_범주,고위험군,고령환자,위험도등급
7,P008,한소희,102,여,160,65,145,95,156,서울대학교병원 노인의학과,True,내원환자,25.4,경도비만,True,True,고위험
8,P009,윤태완,28,남,178,82,118,78,89,서울대학교병원 가정의학과,False,외래환자,25.9,경도비만,False,False,일반
9,P010,송지원,63,여,155,48,120,92,134,서울대학교병원 노인의학과,True,내원환자,20.0,정상체중,False,True,일반


### 🤔 실습 문제

**데이터 삭제 연습**

원본 DataFrame을 보호하면서 불필요한 데이터를 제거하는 실습을 해보세요.

1. **컬럼 삭제**: '분석대상', '데이터출처' 컬럼을 제거한 새로운 DataFrame 생성 (`df_clean` 변수에 할당)
2. **조건부 행 삭제**: BMI가 20 미만인 환자들을 제거한 새로운 DataFrame 생성 (`df_filtered` 변수에 할당)


In [ ]:
# 여기에 작성하세요


In [ ]:
# @title 정답

# 1. 특정 컬럼들 삭제 (원본 보존)
df_clean = df.drop(["분석대상", "데이터출처"], axis=1)
print(f"원본 컬럼 수: {len(df.columns)}개")
print(f"삭제 후 컬럼 수: {len(df_clean.columns)}개")

print("=" * 30)

# 2. 조건부 행 삭제 - BMI 20 미만 환자 제거 (원본 보존)
underweight_patients = df["BMI"] < 22
df_filtered = df.drop(df[underweight_patients].index)
print(f"원본 환자 수: {len(df)}명")
print(f"BMI 20 미만 환자 제거 후: {len(df_filtered)}명")

print("=" * 30)

# 삭제된 환자 확인
print("제거된 환자:")
print(df[underweight_patients][["이름", "BMI", "BMI_범주"]])

print("=" * 30)

# 최종 결과 확인
print("정리된 데이터:")
print(f"df_clean: {df_clean.shape} (컬럼 정리)")
print(f"df_filtered: {df_filtered.shape} (환자 행 필터링)")

원본 컬럼 수: 19개
삭제 후 컬럼 수: 17개
원본 환자 수: 11명
BMI 20 미만 환자 제거 후: 9명
제거된 환자:
    이름   BMI BMI_범주
3  최지은  20.8   정상체중
9  송지원  20.0   정상체중
정리된 데이터:
df_clean: (11, 17) (컬럼 정리)
df_filtered: (9, 19) (환자 행 필터링)
